<launch>
  <!-- size of map, change the size in x, y, z according to your application -->
  <arg name="is_launch_rviz" default="true"/>
  <arg name="use_real_env" value="false"/>
  <arg name="use_fuel_planner" value="false"/>
  <arg name="use_init_motion" value="false"/>
  <arg name="planner_dim" value="3"/>
  
  <arg name="map_size_x" value="320"/>
  <arg name="map_size_y" value="500"/>
  <arg name="map_size_z" value="150"/>

  <arg name="sdf_resolution" value="0.5"/>
  <arg name="sdf_max_ray_length" value="15.0"/>

  <arg name="box_min_x" value="-160.0"/>
  <arg name="box_min_y" value="-250.0"/>
  <arg name="box_min_z" value="-50.0"/>
  <arg name="box_max_x" value="150.0"/>
  <arg name="box_max_y" value="0.0"/>
  <arg name="box_max_z" value="80.0"/>

  <arg name="init_x" value="25.0"/>
  <arg name="init_y" value="-215.0"/>
  <arg name="init_z" value="-20.0"/> 

  <arg name="max_vel" value="3.0" />
  <arg name="max_acc" value="2.0" />

  <!-- topic of your odometry such as VIO or LIO -->
  <arg name="odom_topic" value="/quad_0/lidar_slam/odom" />

  <!-- main algorithm params -->
  <include file="$(find exploration_manager)/launch/sim_virtual_cave/algorithm_virtual_cave.xml">

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>

    <arg name="sdf_resolution" value="$(arg sdf_resolution)"/>
    <arg name="sdf_max_ray_length" value="$(arg sdf_max_ray_length)"/>

    <arg name="use_fuel_planner" value="$(arg use_fuel_planner)"/>
    <arg name="use_real_env" value="$(arg use_real_env)"/>
    <arg name="use_init_motion" value="$(arg use_init_motion)"/>
    <arg name="planner_dim" value="$(arg planner_dim)"/>
    
    <arg name="init_x" value="$(arg init_x)"/>
    <arg name="init_y" value="$(arg init_y)"/>
    <arg name="init_z" value="$(arg init_z)"/>

    <arg name="box_min_x" value="$(arg box_min_x)"/>
    <arg name="box_min_y" value="$(arg box_min_y)"/>
    <arg name="box_min_z" value="$(arg box_min_z)"/>
    <arg name="box_max_x" value="$(arg box_max_x)"/>
    <arg name="box_max_y" value="$(arg box_max_y)"/>
    <arg name="box_max_z" value="$(arg box_max_z)"/>

    <arg name="odometry_topic" value="$(arg odom_topic)"/>
    <!-- sensor pose: transform of camera frame in the world frame -->
    <arg name="sensor_pose_topic" value="/quad0_pcl_render_node/sensor_pose"/>
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- cloud topic: point cloud measurement  -->
    <!-- subscribe ONLY TO ONE of the two topics -->
    <arg name="depth_topic" value="/pcl_render_node/depth"/>
    <arg name="cloud_topic" value="/quad0_pcl_render_node/cloud_all"/>
    <!-- <arg name="cloud_topic" value="/quad0_pcl_render_node/cloud"/> -->


    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="321.04638671875"/>
    <arg name="cy" value="243.44969177246094"/>
    <arg name="fx" value="387.229248046875"/>
    <arg name="fy" value="387.229248046875"/>

    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="$(arg max_vel)" />
    <arg name="max_acc" value="$(arg max_acc)" />
  </include>

  <!-- trajectory server -->
  <!-- <node pkg="plan_manage" name="traj_server" type="traj_server">

    <remap from="/position_cmd" to="planning/pos_cmd"/>

    <remap from="/odom_world" to="$(arg odom_topic)"/>
    <param name="traj_server/time_forward" value="1.5" type="double"/>
    <param name="traj_server/pub_traj_id" value="4" type="int"/>
    <param name="traj_server/init_x" value="$(arg init_x)" type="double"/>
    <param name="traj_server/init_y" value="$(arg init_y)" type="double"/>
    <param name="traj_server/init_z" value="$(arg init_z)" type="double"/>

    <param name="perception_utils/top_angle" value="0.56125" type="double"/>
    <param name="perception_utils/left_angle" value="0.69222" type="double"/>
    <param name="perception_utils/right_angle" value="0.68901" type="double"/>
    <param name="perception_utils/max_dist" value="4.5" type="double"/>
    <param name="perception_utils/vis_dist" value="1.0" type="double"/>

  </node> -->

  <!-- <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen"> -->
  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator">
    <remap from="~odom" to="$(arg odom_topic)"/>        
    <remap from="~goal" to="/move_base_simple/goal"/>
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="point"/>    
  </node>

  <include file="$(find test_interface)/launch/single_avia_with_fuel_virtual_cave.launch">
    <arg name="init_x" value="$(arg init_x)"/>
    <arg name="init_y" value="$(arg init_y)"/>
    <arg name="init_z" value="$(arg init_z)"/>
  </include>

  <include file="$(find fael_exploration)/launch/explorer.launch">
    <arg name="init_x" value="$(arg init_x)"/>
    <arg name="init_y" value="$(arg init_y)"/>
    <arg name="init_z" value="$(arg init_z)"/>
    <arg name="max_vel" value="$(arg max_vel)" />
    <arg name="odometry_topic" value="$(arg odom_topic)"/>
  </include>
  
  <group if="$(arg is_launch_rviz)">
    <include file="$(find exploration_manager)/launch/rviz_sim.launch"/>
  </group>

</launch>
